feather/common/src/block.rs

```diff
@@ -0,0 +1,595 @@
+use crate::entities::player::HotbarSlot;
+use crate::events::BlockChangeEvent;
+use crate::{Game, World};
+use base::{
+    Area, BlockId, BlockPosition, Face, FacingCardinal, FacingCardinalAndDown, FacingCubic,
+    Gamemode, HalfTopBottom, HalfUpperLower, Hinge, Inventory, Item, ItemStack, Position,
+    SimplifiedBlockKind, SlabKind, Vec3d,
+};
+use blocks::BlockKind;
+use ecs::{Ecs, SysResult, SystemExecutor};
+use libcraft_core::{BlockFace, EntityKind};
+use libcraft_items::InventorySlot;
+use quill_common::components::CanBuild;
+use quill_common::events::BlockPlacementEvent;
+use vek::Rect3;
+
+pub fn register(systems: &mut SystemExecutor<Game>) {
+    systems.add_system(block_placement);
+}
+
+/// A system that handles block placement events.
+pub fn block_placement(game: &mut Game) -> SysResult {
+    let mut events = vec![];
+    for (player, event) in game.ecs.query::<&BlockPlacementEvent>().iter() {
+        // Get inventory, gamemode, player position and held item
+        let inv = game.ecs.get_mut::<Inventory>(player)?;
+        let gamemode = game.ecs.get::<Gamemode>(player)?;
+        let hotbar = game.ecs.get::<HotbarSlot>(player)?;
+        let pos = game.ecs.get::<Position>(player)?;
+        let mut slot = match event.hand {
+            libcraft_core::Hand::Main => inv.item(Area::Hotbar, hotbar.get()),
+            libcraft_core::Hand::Offhand => inv.item(Area::Offhand, 0),
+        }
+        .unwrap();
+        // Check whether player has an item in hand, can build and that the item is placeable
+        if slot.is_empty() || !game.ecs.get::<CanBuild>(player)?.0 {
+            continue;
+        }
+        let block = match item_to_block(slot.item_kind().unwrap()) {
+            Some(s) => s,
+            None => continue,
+        };
+        if let Some(mut produced_events) = place_block(
+            &mut game.world,
+            *pos,
+            &game.chunk_entities,
+            block,
+            event,
+            &game.ecs,
+            15,
+        ) {
+            match *gamemode {
+                Gamemode::Survival | Gamemode::Adventure => decrease_slot(&mut slot),
+                _ => {}
+            }
+            events.append(&mut produced_events);
+        }
+    }
+    for e in events {
+        game.ecs.insert_event(e);
+    }
+
+    Ok(())
+}
+/// Blocks get changed if they are getting waterlogged or when they are slabs turning into full blocks.
+/// Blocks get replaced in-place when possible, while changing an adjacent block has a lower priority.
+fn place_block(
+    world: &mut World,
+    player_pos: Position,
+    chunk_entities: &ChunkEntities,
+    block: BlockId,
+    placement: &BlockPlacementEvent,
+    ecs: &Ecs,
+    light_level: u8,
+) -> Option<Vec<BlockChangeEvent>> {
+    let target1 = placement.location;
+    let target_block1 = world.block_at(target1.try_into().unwrap())?;
+    if target_block1.is_air() {
+        return None;
+    }
+    let mut block = block;
```

Comment:
Why not `mut block: BlockId` in a function signature?

Reply:
True.